tutor/plugins/indexes.py

```diff
@@ -0,0 +1,238 @@
+from tutor.exceptions import TutorError
+from tutor.types import Config, get_typed
+
+PLUGIN_INDEXES_KEY = "PLUGIN_INDEXES"
+# Current release name ('zebulon' or 'nightly') and version (1-26)
+RELEASE = __version_suffix__ or env.get_current_open_edx_release_name()
+MAJOR_VERSION = int(__version__.split(".", maxsplit=1)[0])
+
+
+class Indexes:
+    # Store index cache path in this singleton.
+    CACHE_PATH = ""
+
+
+@hooks.Actions.PROJECT_ROOT_READY.add()
+def _set_indexes_cache_path(root: str) -> None:
+    Indexes.CACHE_PATH = env.pathjoin(root, "plugins", "index", "cache.yml")
+
+
+@hooks.Filters.PLUGIN_INDEX_URL.add()
+def _get_index_url_from_alias(url: str) -> str:
+    known_aliases = {
+        "main": "https://overhang.io/tutor/main",
+        "contrib": "https://overhang.io/tutor/contrib",
+        "wizard": "https://overhang.io/tutor/wizard",
+    }
+    return known_aliases.get(url, url)
+
+
+class IndexEntry:
+    def __init__(self, data: dict[str, str]):
+        self._data = data
+
+    @property
+    def data(self) -> dict[str, str]:
+        return self._data
+
+    @property
+    def name(self) -> str:
+        return self.data["name"]
+
+    @property
+    def src(self) -> str:
+        return self.data["src"]
+
+    @property
+    def short_description(self) -> str:
+        lines = self.description.splitlines() or [""]
+        return lines[0][:128]
+
+    @property
+    def description(self) -> str:
+        return self.data.get("description", "").strip()
+
+    @property
+    def author(self) -> str:
+        return self.data.get("author", "")
+
+    @property
+    def maintainer(self) -> str:
+        return self.data.get("maintainer", "")
+
+    @property
+    def url(self) -> str:
+        return self.data.get("url", "")
+
+    @property
+    def index(self) -> str:
+        return self.data["index"]
+
+    def match(self, pattern: str) -> bool:
+        """
+        Simple case-insensitive pattern matching.
+
+        Pattern matching is case-insensitive. Both the name and description fields are
+        searched.
+        """
+        if not pattern:
+            return True
+        pattern = pattern.lower()
+        if pattern in self.name.lower() or pattern in self.description.lower():
+            return True
+        return False
+
+
+def add(url: str, config: Config) -> bool:
+    """
+    Append an index to the list if not already present.
+
+    Return True if if the list of indexes was modified.
+    """
+    indexes = get_all(config)
+    url = hooks.Filters.PLUGIN_INDEX_URL.apply(url)
+    if url in indexes:
+        return False
+    indexes.append(url)
+    return True
+
+
+def remove(url: str, config: Config) -> bool:
+    """
+    Remove an index to the list if present.
+
+    Return True if if the list of indexes was modified.
+    """
+    indexes = get_all(config)
+    url = hooks.Filters.PLUGIN_INDEX_URL.apply(url)
+    if url not in indexes:
+        return False
+    indexes.remove(url)
```

Comment:
What happens if somebody removes one of the official indices?

Reply:
It's just like removing an apt or pypi index: nothing much happens. Installed plugins are not uninstalled or deactivated. But plugins will no longer be automatically upgraded with `tutor plugins upgrade all` and cannot be installed or uninstalled.

Reply:
Makes sense, thanks!